File: LeetCode/104. Maximum Depth of Binary Tree/elixir/solution.exs
# Definition for a binary tree node.
#
# defmodule TreeNode do
#   @type t :: %__MODULE__{
#           val: integer,
#           left: TreeNode.t() | nil,
#           right: TreeNode.t() | nil
#         }
#   defstruct val: 0, left: nil, right: nil
# end

defmodule Solution do
  @spec max_depth(root :: TreeNode.t() | nil) :: integer
  def max_depth(root) do
    get_depth(root)
  end

  defp get_depth(node, depth \\ 0)

  defp get_depth(nil, depth) do
    depth
  end

  defp get_depth(%{left: l_node, right: r_node}, depth) do
    max(get_depth(l_node, depth + 1), get_depth(r_node, depth + 1))
  end
end

File: LeetCode/104. Maximum Depth of Binary Tree/java/Solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public int maxDepth(TreeNode root) {
        return getDepth(root, 0);
    }

    public int getDepth(TreeNode node, int deps) {
        if (node == null) {
            return deps;
        }
        return Math.max(getDepth(node.left, deps + 1), getDepth(node.right, deps + 1));
    }
}
